source/Orders.cpp

```diff
@@ -0,0 +1,333 @@
+{
+	return targetShip.lock();
+}
+
+
+
+void Orders::SetTargetSystem(const System *system)
+{
+	targetSystem = system;
+}
+
+
+
+shared_ptr<Minable> Orders::GetTargetAsteroid() const
+{
+	return targetAsteroid.lock();
+}
+
+
+
+const Point &Orders::GetTargetPoint() const
+{
+	return targetPoint;
+}
+
+
+
+const System *Orders::GetTargetSystem() const
+{
+	return targetSystem;
+}
+
+
+
+void Orders::UpdateOrder(const Ship *orderedShip, const System *flagshipSystem)
+{
+	if(HasMine() && orderedShip->Cargo().Free() && targetAsteroid.expired())
+		SetHarvest();
+	else if((activeOrders & REQUIRES_TARGET).any())
+	{
+		shared_ptr<Ship> ship = GetTargetShip();
+		shared_ptr<Minable> asteroid = GetTargetAsteroid();
+		// Check if the target ship itself is targetable, or if it is one of your ships that you targeted.
+		bool invalidTarget = !ship
+				|| (!ship->IsTargetable() && orderedShip->GetGovernment() != ship->GetGovernment())
+				|| (ship->IsDisabled() && HasAttack());
+		// Alternatively, if an asteroid is targeted, then it is not an invalid target.
+		invalidTarget &= !asteroid;
+
+		// Check if the target ship is in a system where we can target.
+		// This check only checks for undocked ships (that have a current system).
+		const System *orderSystem = orderedShip->GetSystem();
+		const System *shipSystem = ship ? ship->GetSystem() : nullptr;
+		bool targetOutOfReach = !ship || (orderSystem && shipSystem != orderSystem
+				&& shipSystem != flagshipSystem);
+		// Asteroids are never out of reach since they're in the same system as the flagship.
+		targetOutOfReach &= !asteroid;
+
+		// Cancel any orders that required a target.
+		if(invalidTarget || targetOutOfReach)
+			activeOrders &= ~REQUIRES_TARGET;
+	}
+}
+
+
+
+void Orders::MergeOrders(const Orders &other, bool &hasMismatch, bool &alreadyHarvesting, int &orderOperation)
+{
+	// HOLD_ACTIVE cannot be given as a manual order, but we make sure here
+	// that any HOLD_ACTIVE order also matches when a HOLD_POSITION
+	// command is given.
+	if(HasHoldActive())
+		ApplyOrder(HOLD_POSITION);
+
+	bool newTargetShip = (GetTargetShip() != other.GetTargetShip());
+	bool newTargetAsteroid = (GetTargetAsteroid() != other.GetTargetAsteroid());
+	hasMismatch |= newTargetShip || newTargetAsteroid;
+
+	// For each bit of the other order that is flipped, apply the flip to this order.
+	for(size_t i = 0; i < other.activeOrders.size(); ++i)
+		if(other.activeOrders.test(i))
+		{
+			hasMismatch |= !activeOrders.test(i);
+			// If the existing order had a target and the new order is of the same type
+			// but with a different target, then no change needs to be made to this bit.
+			// Only run this check if the order operation is 2, as that means that this
+			// is the first order that is being evaluated and will set the order
+			// operation for all subsequent orders.
+			OrderType type = static_cast<OrderType>(i);
+			if(orderOperation == 2)
+			{
+				if(((type == OrderType::ATTACK && HasAttack())
```

Comment:
Based on the `REQUIRES_TARGET` list above, is this missing a check for `GATHER`? Also, would it be possible to have `REQUIRES_TARGET` and `REQUIRES_ASTEROID` bitsets you can just test against instead of this long if statement?

Reply:
From what i could tell, GATHER can only ever target the player's flagship. so it seemed like there'd never be a point where a GATHER order was sent for a separate ship.

Reply:
Using the REQUIRES_TARGET bitset here covers that, though.